File: Risk/RiskPoint.m
//
// This file is a part of Risk by Mike Ferris.
//

#import "Risk.h"

RCSID ("$Id: RiskPoint.m,v 1.2 1997/12/15 07:44:13 nygard Exp $");

#import "RiskPoint.h"

#if !__has_feature(objc_arc)
#error this file needs to be compiled with Automatic Reference Counting (ARC)
#endif

#define RiskPoint_VERSION 1

@implementation RiskPoint
@synthesize point;

+ (void) initialize
{
    if (self == [RiskPoint class])
    {
        [self setVersion:RiskPoint_VERSION];
    }
}

//----------------------------------------------------------------------

+ (instancetype) riskPointWithPoint:(NSPoint)aPoint
{
    return [[RiskPoint alloc] initWithPoint:aPoint];
}

//----------------------------------------------------------------------

- (instancetype) initWithPoint:(NSPoint)aPoint
{
    if (self = [super init]) {
        point = aPoint;
    }
    
    return self;
}

- (instancetype)init
{
    return [self initWithPoint:NSZeroPoint];
}

//----------------------------------------------------------------------

#define kRiskPoint @"Point"
- (void) encodeWithCoder:(NSCoder *)aCoder
{
    [aCoder encodePoint:point forKey:kRiskPoint];
}

//----------------------------------------------------------------------

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    NSPoint bPoint;
    if (aDecoder.allowsKeyedCoding && [aDecoder containsValueForKey:kRiskPoint]) {
        bPoint = [aDecoder decodePointForKey:kRiskPoint];
    } else {
        bPoint = [aDecoder decodePoint];
    }
    return [self initWithPoint:bPoint];
}

//----------------------------------------------------------------------

- (NSString *) description
{
    return [NSString stringWithFormat:@"<RiskPoint: %f,%f>", point.x, point.y];
}

@end

File: Risk/RiskPoint.swift
//
//  RiskPoint.swift
//  Risk
//
//  Created by C.W. Betts on 6/15/16.
//  Copyright © 2016 C.W. Betts. All rights reserved.
//

import Foundation

private let RiskPoint_VERSION = 1
private let kRiskPoint = "Point"

/// A RiskPoint can be encoded on a stream and stored in arrays.
@objc(RiskPoint) final class RiskPoint: NSObject, NSSecureCoding {
	@objc let point: NSPoint
	private static var __doSomethingOnce: () = {
		RiskPoint.setVersion(RiskPoint_VERSION)
	}()
	
	//----------------------------------------------------------------------
	
	@objc class func setUpVersions() {
		_=RiskPoint.__doSomethingOnce
	}
	
	//----------------------------------------------------------------------
	
	@objc init(point: NSPoint) {
		self.point = point
		super.init()
	}
	
	//----------------------------------------------------------------------
	
	override convenience init() {
		self.init(point: .zero)
	}
	
	//----------------------------------------------------------------------
	
	static var supportsSecureCoding: Bool {
		return true
	}
	
	//----------------------------------------------------------------------
	
	func encode(with aCoder: NSCoder) {
		aCoder.encode(point, forKey: kRiskPoint)
	}
	
	//----------------------------------------------------------------------
	
	convenience init?(coder aDecoder: NSCoder) {
		let bPoint: NSPoint
		if aDecoder.allowsKeyedCoding {
			bPoint = aDecoder.decodePoint(forKey: kRiskPoint)
		} else {
			bPoint = aDecoder.decodePoint()
		}
		self.init(point: bPoint)
	}
	
	//----------------------------------------------------------------------
	
	override var description: String {
		return "<RiskPoint: \(point.x),\(point.y)>"
	}
	
	//----------------------------------------------------------------------
	
}
